pkg/storage/tsdb/config.go

```diff
@@ -209,6 +213,7 @@ func (cfg *TSDBConfig) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.WALCompressionEnabled, "blocks-storage.tsdb.wal-compression-enabled", false, "True to enable TSDB WAL compression.")
 	f.IntVar(&cfg.WALSegmentSizeBytes, "blocks-storage.tsdb.wal-segment-size-bytes", wal.DefaultSegmentSize, "TSDB WAL segments files max size (bytes).")
 	f.BoolVar(&cfg.FlushBlocksOnShutdown, "blocks-storage.tsdb.flush-blocks-on-shutdown", false, "True to flush blocks to storage on shutdown. If false, incomplete blocks will be reused after restart.")
+	f.DurationVar(&cfg.CloseIdleTSDBTimeout, "blocks-storage.tsdb.close-idle-tsdb-timeout", 12*time.Hour, "If TSDB has not received any data for this duration, and all blocks from TSDB have been shipped, TSDB is closed and deleted from local disk. If set to positive value, this value should be equal or higher than -querier.query-ingesters-within flag to make sure that TSDB is not closed prematurely, which could cause invalid query results. 0 or negative value disables closing of idle TSDB.")
```

Comment:
> Is that flag, or `-querier.query-store-after` or both? I think the minimum value is the max between `-querier.query-ingesters-within` and `-querier.query-store-after`.

`-querier.query-store-after` is about querying long-term storage, which in my opinion is unrelated.

Reply:
Yes, you're right.

Reply:
> Also 12h default looks a bit high. The default value for `-blocks-storage.tsdb.head-compaction-idle-timeout`, so I think 2h may be safe for this default. In our OSS jsonnet we'll override it according to our setup config.

On the other hand, `-querier.query-ingesters-within` defaults to 0, so perhaps we should disable closing of idle TSDBs by default?

Reply:
We may want to default it to 0 until we're comfortable it works fine (BTW, we may mark it experimental) but I don't think we should default to 0 because `-querier.query-ingesters-within` is `0`. Blocks are kept in the ingester up to the retention period (not forever), but practically they will be available to query back from store-gateway way before `12h` so I think `12h` is a good default.

Reply:
> practically they will be available to query back from store-gateway way before 12h so I think 12h is a good default.

Fair enough. I will keep it at 0 while it's experimental, and let's use non-zero value once we test that it works as expected.